In [1]:
import sklearn
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('./Downloads/HW1/part1/Linear_models_hw2/weatherAUS.csv')
data

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
0       2008-12-01   Albury     13.4     22.9       0.6          NaN   
1       2008-12-02   Albury      7.4     25.1       0.0          NaN   
2       2008-12-03   Albury     12.9     25.7       0.0          NaN   
3       2008-12-04   Albury      9.2     28.0       0.0          NaN   
4       2008-12-05   Albury     17.5     32.3       1.0          NaN   
...            ...      ...      ...      ...       ...          ...   
142188  2017-06-20    Uluru      3.5     21.8       0.0          NaN   
142189  2017-06-21    Uluru      2.8     23.4       0.0          NaN   
142190  2017-06-22    Uluru      3.6     25.3       0.0          NaN   
142191  2017-06-23    Uluru      5.4     26.9       0.0          NaN   
142192  2017-06-24    Uluru      7.8     27.0       0.0          NaN   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  \
0            NaN           W           44.0          W  ...        22.0   
1            NaN         WNW           44.0        NNW  ...        25.0   
2            NaN         WSW           46.0          W  ...        30.0   
3            NaN          NE           24.0         SE  ...        16.0   
4            NaN           W           41.0        ENE  ...        33.0   
...          ...         ...            ...        ...  ...         ...   
142188       NaN           E           31.0        ESE  ...        27.0   
142189       NaN           E           31.0         SE  ...        24.0   
142190       NaN         NNW           22.0         SE  ...        21.0   
142191       NaN           N           37.0         SE  ...        24.0   
142192       NaN          SE           28.0        SSE  ...        24.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0            1007.7       1007.1       8.0       NaN     16.9     21.8   
1            1010.6       1007.8       NaN       NaN     17.2     24.3   
2            1007.6       1008.7       NaN       2.0     21.0     23.2   
3            1017.6       1012.8       NaN       NaN     18.1     26.5   
4            1010.8       1006.0       7.0       8.0     17.8     29.7   
...             ...          ...       ...       ...      ...      ...   
142188       1024.7       1021.2       NaN       NaN      9.4     20.9   
142189       1024.6       1020.3       NaN       NaN     10.1     22.4   
142190       1023.5       1019.1       NaN       NaN     10.9     24.5   
142191       1021.0       1016.8       NaN       NaN     12.5     26.1   
142192       1019.4       1016.5       3.0       2.0     15.1     26.0   

        RainToday  RISK_MM  RainTomorrow  
0              No      0.0            No  
1              No      0.0            No  
2              No      0.0            No  
3              No      1.0            No  
4              No      0.2            No  
...           ...      ...           ...  
142188         No      0.0            No  
142189         No      0.0            No  
142190         No      0.0            No  
142191         No      0.0            No  
142192         No      0.0            No  

[142193 rows x 24 columns]

In [3]:
target = data['RainTomorrow']
cat_columns = data.select_dtypes(include=['object']).drop(['RainTomorrow'], axis = 1)
num_columns = data.select_dtypes(include=['float64'])

In [4]:
cat_indices = cat_columns.columns
num_indices = num_columns.columns

In [5]:
data = data.drop(cat_indices, axis = 1)

In [6]:
for series in num_indices:
    if data[series].isna().sum()>0:
        data = data.replace({series:{np.nan:data[series].mean()}})
data

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
0          13.4     22.9       0.6     5.469824  7.624853           44.0   
1           7.4     25.1       0.0     5.469824  7.624853           44.0   
2          12.9     25.7       0.0     5.469824  7.624853           46.0   
3           9.2     28.0       0.0     5.469824  7.624853           24.0   
4          17.5     32.3       1.0     5.469824  7.624853           41.0   
...         ...      ...       ...          ...       ...            ...   
142188      3.5     21.8       0.0     5.469824  7.624853           31.0   
142189      2.8     23.4       0.0     5.469824  7.624853           31.0   
142190      3.6     25.3       0.0     5.469824  7.624853           22.0   
142191      5.4     26.9       0.0     5.469824  7.624853           37.0   
142192      7.8     27.0       0.0     5.469824  7.624853           28.0   

        WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0               20.0          24.0         71.0         22.0       1007.7   
1                4.0          22.0         44.0         25.0       1010.6   
2               19.0          26.0         38.0         30.0       1007.6   
3               11.0           9.0         45.0         16.0       1017.6   
4                7.0          20.0         82.0         33.0       1010.8   
...              ...           ...          ...          ...          ...   
142188          15.0          13.0         59.0         27.0       1024.7   
142189          13.0          11.0         51.0         24.0       1024.6   
142190          13.0           9.0         56.0         21.0       1023.5   
142191           9.0           9.0         53.0         24.0       1021.0   
142192          13.0           7.0         51.0         24.0       1019.4   

        Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RISK_MM  \
0            1007.1  8.000000  4.503167     16.9     21.8      0.0   
1            1007.8  4.437189  4.503167     17.2     24.3      0.0   
2            1008.7  4.437189  2.000000     21.0     23.2      0.0   
3            1012.8  4.437189  4.503167     18.1     26.5      1.0   
4            1006.0  7.000000  8.000000     17.8     29.7      0.2   
...             ...       ...       ...      ...      ...      ...   
142188       1021.2  4.437189  4.503167      9.4     20.9      0.0   
142189       1020.3  4.437189  4.503167     10.1     22.4      0.0   
142190       1019.1  4.437189  4.503167     10.9     24.5      0.0   
142191       1016.8  4.437189  4.503167     12.5     26.1      0.0   
142192       1016.5  3.000000  2.000000     15.1     26.0      0.0   

       RainTomorrow  
0                No  
1                No  
2                No  
3                No  
4                No  
...             ...  
142188           No  
142189           No  
142190           No  
142191           No  
142192           No  

[142193 rows x 18 columns]

In [7]:
#from sklearn.preprocessing import StandardScaler
#scaler = StandardScaler(copy = False)
#numeric_features = pd.DataFrame(scaler.fit_transform(data[num_indices]), columns = [num_indices])
#numeric_features

In [8]:
#data = pd.concat([numeric_features, target], axis=1, join = 'inner')
data.replace({'RainTomorrow':{'No': 0, 'Yes' : 1}}, inplace = True)

In [9]:
data['RainTomorrow'].value_counts()

0    110316
1     31877
Name: RainTomorrow, dtype: int64

In [10]:
data

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
0          13.4     22.9       0.6     5.469824  7.624853           44.0   
1           7.4     25.1       0.0     5.469824  7.624853           44.0   
2          12.9     25.7       0.0     5.469824  7.624853           46.0   
3           9.2     28.0       0.0     5.469824  7.624853           24.0   
4          17.5     32.3       1.0     5.469824  7.624853           41.0   
...         ...      ...       ...          ...       ...            ...   
142188      3.5     21.8       0.0     5.469824  7.624853           31.0   
142189      2.8     23.4       0.0     5.469824  7.624853           31.0   
142190      3.6     25.3       0.0     5.469824  7.624853           22.0   
142191      5.4     26.9       0.0     5.469824  7.624853           37.0   
142192      7.8     27.0       0.0     5.469824  7.624853           28.0   

        WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0               20.0          24.0         71.0         22.0       1007.7   
1                4.0          22.0         44.0         25.0       1010.6   
2               19.0          26.0         38.0         30.0       1007.6   
3               11.0           9.0         45.0         16.0       1017.6   
4                7.0          20.0         82.0         33.0       1010.8   
...              ...           ...          ...          ...          ...   
142188          15.0          13.0         59.0         27.0       1024.7   
142189          13.0          11.0         51.0         24.0       1024.6   
142190          13.0           9.0         56.0         21.0       1023.5   
142191           9.0           9.0         53.0         24.0       1021.0   
142192          13.0           7.0         51.0         24.0       1019.4   

        Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RISK_MM  \
0            1007.1  8.000000  4.503167     16.9     21.8      0.0   
1            1007.8  4.437189  4.503167     17.2     24.3      0.0   
2            1008.7  4.437189  2.000000     21.0     23.2      0.0   
3            1012.8  4.437189  4.503167     18.1     26.5      1.0   
4            1006.0  7.000000  8.000000     17.8     29.7      0.2   
...             ...       ...       ...      ...      ...      ...   
142188       1021.2  4.437189  4.503167      9.4     20.9      0.0   
142189       1020.3  4.437189  4.503167     10.1     22.4      0.0   
142190       1019.1  4.437189  4.503167     10.9     24.5      0.0   
142191       1016.8  4.437189  4.503167     12.5     26.1      0.0   
142192       1016.5  3.000000  2.000000     15.1     26.0      0.0   

        RainTomorrow  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
...              ...  
142188             0  
142189             0  
142190             0  
142191             0  
142192             0  

[142193 rows x 18 columns]

In [11]:
X = data.drop(['RainTomorrow'], axis = 1)
y = data['RainTomorrow']

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=24, shuffle = False)

In [13]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler(copy = False)
X_train = pd.DataFrame(scaler.fit_transform(X_train), columns = [num_indices])
X_train

MinTemp   MaxTemp  Rainfall Evaporation  Sunshine WindGustSpeed  \
0       0.226229  0.023195 -0.207679    0.023220  0.045290      0.311176   
1      -0.719958  0.340463 -0.276485    0.023220  0.045290      0.311176   
2       0.147380  0.426991 -0.276485    0.023220  0.045290      0.459672   
3      -0.436102  0.758681 -0.276485    0.023220  0.045290     -1.173789   
4       0.872790  1.378796 -0.161808    0.023220  0.045290      0.088431   
...          ...       ...       ...         ...       ...           ...   
106639 -0.278404 -0.669027 -0.276485   -0.363058  0.291534      0.013016   
106640 -0.073397 -0.928611  0.617993   -0.910616 -2.443913      0.013016   
106641  0.084301 -0.712291 -0.276485   -0.971455 -1.677988      0.013016   
106642  0.084301 -0.943032 -0.207679   -0.728096  0.984514      0.013016   
106643 -0.719958 -0.409444 -0.276485   -0.545577  1.495131      0.013016   

       WindSpeed9am WindSpeed3pm Humidity9am Humidity3pm Pressure9am  \
0          0.668983     0.574470    0.032980   -1.500200   -1.537524   
1         -1.093498     0.354925   -1.455997   -1.353463   -1.104936   
2          0.558828     0.794014   -1.786881   -1.108901   -1.552441   
3         -0.322413    -1.072111   -1.400850   -1.793674   -0.060759   
4         -0.763033     0.135381    0.639601   -0.962164   -1.075102   
...             ...          ...         ...         ...         ...   
106639    -0.542723     1.342874   -0.022167    0.798680    0.386745   
106640    -0.763033    -0.633023    1.025632    1.043242    0.908834   
106641     0.668983     2.989455   -0.077314    0.749768    0.267411   
106642    -1.093498    -0.633023   -1.180261    0.016083    1.938094   
106643     0.668983     1.562418   -0.573640    0.358469    1.953011   

       Pressure3pm  Cloud9am  Cloud3pm   Temp9am   Temp3pm   RISK_MM  
0        -1.293646  1.513260 -0.044422  0.056915  0.080648 -0.275889  
1        -1.187810 -0.044604 -0.044422  0.104288  0.452960 -0.275889  
2        -1.051736 -0.044604 -1.228171  0.704345  0.289143 -0.275889  
3        -0.431841 -0.044604 -0.044422  0.246407  0.780595 -0.161978  
4        -1.459960  1.076003  1.609233  0.199034  1.257154 -0.253107  
...            ...       ...       ...       ...       ...       ...  
106639    0.520681  0.638746  0.190531 -0.100995 -0.634191  0.612619  
106640    0.974263  1.513260  0.663431 -0.401024 -0.857578 -0.275889  
106641    1.034741  0.638746  1.609233 -0.164159 -0.932040 -0.207543  
106642    2.123338 -0.235768 -2.173973 -0.243114 -0.842686 -0.275889  
106643    1.548800 -1.984796 -2.173973 -0.243114 -0.485266 -0.275889  

[106644 rows x 17 columns]

In [14]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler(copy = False)
X_test = pd.DataFrame(scaler.fit_transform(X_test), columns = [num_indices])
X_test

MinTemp   MaxTemp  Rainfall Evaporation  Sunshine WindGustSpeed  \
0     -0.470224  0.486475 -0.290522    0.181107  1.269675     -0.043539   
1     -0.254333  0.365206 -0.290522    0.036564  1.269675     -0.043539   
2      0.100343 -0.928325 -0.210079    0.470195 -1.864746     -0.043539   
3     -0.763217 -1.305605  2.471358   -0.975243 -0.110963     -0.043539   
4     -0.932845 -0.982222 -0.290522   -0.902972  1.083103     -0.043539   
...         ...       ...       ...         ...       ...           ...   
35544 -1.441730 -0.389354 -0.290522   -0.082749 -0.139003     -0.787319   
35545 -1.549675 -0.173765 -0.290522   -0.082749 -0.139003     -0.787319   
35546 -1.426309  0.082246 -0.290522   -0.082749 -0.139003     -1.532399   
35547 -1.148736  0.297835 -0.290522   -0.082749 -0.139003     -0.290598   
35548 -0.778638  0.311309 -0.290522   -0.082749 -0.139003     -1.035679   

      WindSpeed9am WindSpeed3pm Humidity9am Humidity3pm Pressure9am  \
0        -1.012239     0.101015    0.187333   -0.339401    0.784838   
1        -1.012239     2.526029   -0.106100    1.427051   -1.285625   
2        -1.258309     2.526029    0.285144    2.015869   -1.448304   
3         0.587216    -0.168431   -0.203911    0.200348    0.873572   
4        -0.889204    -0.437877   -0.350627    0.543825    1.494711   
...            ...          ...         ...         ...         ...   
35544     0.095076    -0.707323   -0.252816   -1.026355    1.198930   
35545    -0.150994    -0.976768   -0.644060   -1.173560    1.184141   
35546    -0.150994    -1.246214   -0.399533   -1.320764    1.021462   
35547    -0.643134    -1.246214   -0.546249   -1.173560    0.651737   
35548    -0.150994    -1.515660   -0.644060   -1.173560    0.415112   

      Pressure3pm  Cloud9am  Cloud3pm   Temp9am   Temp3pm   RISK_MM  
0        0.331643 -1.858194 -2.060190 -0.482567  0.143237 -0.290225  
1       -1.463220 -1.408396  0.380001  0.249212 -0.414926 -0.210230  
2       -0.751208  0.840594  1.844116 -0.198816 -1.419619  2.456259  
3        1.325493  0.840594 -1.572152 -0.840989 -1.182399 -0.290225  
4        1.266159  0.840594 -1.572152 -0.557238 -0.889364 -0.290225  
...           ...       ...       ...       ...       ...       ...  
35544    1.058489  0.137645  0.137528 -1.333819 -0.289339 -0.290225  
35545    0.924987  0.137645  0.137528 -1.229279 -0.080028 -0.290225  
35546    0.746984  0.137645  0.137528 -1.109805  0.213007 -0.290225  
35547    0.405811  0.137645  0.137528 -0.870857  0.436272 -0.290225  
35548    0.361311 -0.508800 -1.084113 -0.482567  0.422318 -0.290225  

[35549 rows x 17 columns]

Ближайшие соседи

In [29]:
y_train = np.array(y_train).reshape(-1,)
y_test = np.array(y_test).reshape(-1,)

In [30]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
model1 = KNeighborsClassifier()
gcv = GridSearchCV(estimator=model1,
                  cv = 3,
                  param_grid={'n_neighbors': [5,10,15,20]},
                  scoring='accuracy',
                  refit = True,
                  n_jobs = -1,
                  )

gcv.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=KNeighborsClassifier(), n_jobs=-1,
             param_grid={'n_neighbors': [5, 10, 15, 20]}, scoring='accuracy')

In [33]:
gcv.best_params_

{'n_neighbors': 15}

In [34]:
y_predicted =  gcv.best_estimator_.predict(X_test)
accuracy = np.sum(y_predicted==y_test)/len(y_test)

In [35]:
accuracy

0.8980280739261302

In [36]:
from sklearn.metrics import accuracy_score
accuracy_score(y_pred, y_test)

0.9591268390109426

In [37]:
%%time
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier(n_neighbors = 15)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
accuracy_score(y_pred, y_test)

0.8980280739261302

In [38]:
from sklearn.metrics import roc_auc_score
roc_auc_score(y_test, y_pred)

0.7929119753562421

The best kneighbours classificator:
params: {'n_neighbors': 15}
time to fit and predict: 5min 6s
accuracy: 0.8980280739261302
roc_auc_score: 0.7929119753562421

Логистическая регрессия

In [15]:
import numpy as np

In [16]:
class Logistic_Regression:
    def __init__(self, l1_ratio):
        self.l1_ratio = l1_ratio
        self.w = None
    def generate_batches(self, X, y, batch_size):
        np.random.seed(42)
        X = np.array(X)
        y = np.array(y)
        perm = np.random.permutation(len(X))
        for n in range(0,len(X)//batch_size):
            yield (X[perm][n*batch_size:n*batch_size+batch_size],y[perm][n*batch_size:n*batch_size+batch_size])
    
    def fit(self, X, y, num_epochs = 5, lr = 0.03, batch_size = 1000):
        if self.w is None:
            np.random.seed(42)
            self.w = np.random.randn(X.shape[1] + 1)
        
        X_train = np.concatenate((np.ones((X.shape[0], 1)), X), axis=1)
        
        losses = []
        
        for i in range(num_epochs):
            for X_batch, y_batch in self.generate_batches(X_train, y, batch_size):
                predictions = self.sigmoid(X_batch) 
                predictions = np.clip(predictions, 1e-10, 1 - 1e-10)
                losses.append(-np.mean(y_batch * np.log(predictions) + (1 - y_batch) * np.log(1 - predictions)))
                self.w -= self.get_grad(X_batch, y_batch) * lr
        return losses

    def predict(self, X, treshold = 0.5):
        X_ = np.concatenate((np.ones((X.shape[0], 1)), X), axis=1)
        return self.sigmoid(X_)>treshold
    
    def get_grad(self, X, y):
        grad_basic = np.dot(X.T, (self.sigmoid(X) - y))
        grad_l1 =  self.l1_ratio * self.w
        grad_l2 = 1/2*(1-self.l1_ratio) * np.sign(self.w)
        return grad_basic + grad_l1 + grad_l2
    
    def sigmoid(self, X):
        return 1/(1+np.exp((-1)*np.dot(X,self.w)))                    

In [19]:
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import r2_score
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
import itertools

In [22]:
%%time
for l1r,bs,lr_ in itertools.product([0.2,0.5,0.8],[5000,10000,15000],[0.001,0.03,0.1]):
    model = Logistic_Regression(l1r)
    losses = model.fit(X_train, y_train, lr = lr_, batch_size = bs)
    
    y_pred = model.predict(X_test)
    print("params: ", l1r,bs,lr_, "roc_auc: ", roc_auc_score(y_pred, y_test), 
          "accuracy: ", accuracy_score(y_test, y_pred), "r2_score:", r2_score(y_test, y_pred), "f1_score: ", f1_score(y_test, y_pred),
          "recall: ", recall_score(y_test, y_pred))

params:  0.2 5000 0.001 roc_auc:  0.9785747921985544 accuracy:  0.9688317533545248 r2_score: 0.8157456445017356 f1_score:  0.9225391498881432 recall:  0.8605712795095866
params:  0.2 5000 0.03 roc_auc:  0.9763226795265582 accuracy:  0.9782272356465723 r2_score: 0.8712880224226925 f1_score:  0.9482343499197432 recall:  0.9246119733924612
params:  0.2 5000 0.1 roc_auc:  0.8312258104828006 accuracy:  0.8767053925567526 r2_score: 0.27113101069594503 f1_score:  0.6860989758647855 recall:  0.6247554454154167
params:  0.2 10000 0.001 roc_auc:  0.971025872861898 accuracy:  0.9655123913471546 r2_score: 0.7961228882302598 f1_score:  0.9147308387814717 recall:  0.857701839050476
params:  0.2 10000 0.03 roc_auc:  0.6392803450663561 accuracy:  0.7088244395060339 r2_score: -0.7213148319156453 f1_score:  0.4923740866068364 recall:  0.654754141124299
params:  0.2 10000 0.1 roc_auc:  0.9445027885177316 accuracy:  0.9511941264170581 r2_score: 0.7114789649914361 f1_score:  0.8804025642793134 recall:  0.8

In [23]:
%%time
model = Logistic_Regression(0.2)
model.fit(X_train, y_train, lr = 0.03, batch_size = 5000)
y_pred = (model.predict(X_test)>0.5)
roc_auc_score(y_pred, y_test),accuracy_score(y_test, y_pred),r2_score(y_test, y_pred)


Wall time: 7.43 s


(0.9763226795265582, 0.9782272356465723, 0.8712880224226925)

The best logistic regression classificator:
params: {'l1_ratio' : 0.2, 'batch_size':5000, 'lr' : 0.03}
time to fit and predict: 2.9 s
accuracy: 0.9782272356465723
roc_auc_score: 0.9763226795265582

Наивный Байес

In [24]:
from sklearn.naive_bayes import  BernoulliNB
from sklearn.naive_bayes import  MultinomialNB
from sklearn.naive_bayes import  GaussianNB

In [25]:
model_bnb = BernoulliNB()
model_mnb = MultinomialNB()
model_gnb = GaussianNB()

In [26]:
%%time
model_bnb.fit(X_train, y_train)
y_pred = model_bnb.predict(X_test)
accuracy_score(y_pred, y_test), roc_auc_score(y_pred, y_test)

Wall time: 244 ms


(0.8618807842695997, 0.7975839199221166)

In [27]:
%%time
model_mnb.fit(np.abs(X_train), y_train)
y_pred = model_mnb.predict(np.abs(X_test))
accuracy_score(y_pred, y_test), roc_auc_score(y_pred, y_test)

Wall time: 185 ms


(0.8330754732903879, 0.8760253352296776)

In [28]:
%%time
model_gnb.fit(X_train, y_train)
y_pred = model_gnb.predict(X_test)
accuracy_score(y_pred, y_test), roc_auc_score(y_pred, y_test)

Wall time: 255 ms


(0.9591268390109426, 0.9203583585377758)

The best NB classificator:
type: GaussianNB
time to fit and predict: 310 ms
accuracy: 0.956313820360629
roc_auc_score: 0.9181146819191086

Cравнив результаты классификаторам по метрикам accuracy, roc_auc и по времени обучения и предсказания можно заключить, что Logistic Regression - самый лучший классиффикатор для данной задачи c accuracy: 0.9782272356465723 и time: 2.9 s